compiler/backend/src/org/jetbrains/kotlin/codegen/optimization/common/OptimizationBasicInterpreter.java

```diff
@@ -380,4 +378,34 @@ public BasicValue merge(
     private static boolean isReference(@NotNull BasicValue v) {
         return v.getType().getSort() == Type.OBJECT || v.getType().getSort() == Type.ARRAY;
     }
+
+    // Merge reference types, keeping track of array dimensions.
+    // See also org.jetbrains.org.objectweb.asm.Frame.merge.
+    private BasicValue mergeReferenceTypes(@NotNull Type a, @NotNull Type b) {
```

Comment:
Though it does not shoot now due to check above (`if (v.equals(w)) return v;`), wouldn't it be better to handle case `a == b`? It might be strange that merging `[I` and `[I` gives `Object`

Reply:
I'd like to avoid any unneeded checks, since this place _might_ be a hotspot in some scenarios. So I can probably add the info that types should not be equal at this point to the method's comment, what do you think?

Reply:
Comment would be ok I guess.